crates/runner/src/memory/mod.rs

```diff
@@ -264,32 +302,61 @@ impl Memory {
         }
 
         // Check that the entire slice fits within memory limits
-        let start_address = start_addr.0 as usize;
         let slice_len = values.len();
         // Since we already checked for empty slice, slice_len >= 1
         let last_addr = start_addr.0.checked_add((slice_len - 1) as u32).ok_or(
             MemoryError::AddressOutOfBounds {
                 addr: start_addr,
-                max_addr: 1 << MAX_MEMORY_SIZE_BITS,
+                max_addr: MAX_ADDRESS as u32,
```

Comment:
maybe out of scope of this PR but `AddressOutOfBounds` could be simplified to remove the `max_addr` as it will be always `MAX_ADDRESS`